### ISSUE TITLE
Bug in Tm_NN in MeltingTemp?

### ISSUE THREAD

**Issue by larssnip** [biopython/biopython#1051]
I observe some really crazy melting temperatures output from the Tm_NN Method in MeltingTemp (Bio.SeqUtils). First, the sensible result:

>>> MeltingTemp.Tm_NN("GGCCACGCCGA", dnac1=200,Na=0,Mg=4)
57.12...  (which is probably Close to the truth)

Then, the crazy result:

>>> MeltingTemp.Tm_NN("GGCCACGCCGA", c_seq="CAGCTACTGAT", dnac1=200,Na=0,Mg=4)
13078.65... (which is NOT sensible at all)

Notice there is match and mismatch at every second position in the duplex. A small change to dnac1 also makes the temperature change sign, but keep the horrible absolute value! It indicates a case of "almost dividing by 0"... Is this because the Delta_S becomes Close to zero? There should be some guarding against this, I think...

**Comment by peterjc** [comment]
It probably doesn't make much difference, but which version of Biopython do you have.

Do you think the could could do something like issue a warning if over half the base pairs don't match? e.g.

``` python
import warnings
from Bio import BiopythonWarning
...
warnings.warn("Melting temperature will unreliable as the sequences are very different", BiopythonWarning)
```

**Comment by larssnip** [comment]
Yes, a warning or something like that is needed here, I think.

It is also amazing how little is known about melting temperatures in general. The IMM-table contains Delta-H and Delta-H values for only 3 double mismatches, and I suppose this is all that is known (in the entire world) about this? I haven't found anything better than BioPython in thsi direction. Given the HUGE amount of PCR-like Technology being used every single day it is really remarkable that we have Close to no idea how well it Works (or not)...!

**Comment by MarkusPiotrowski** [comment]
I will have a look if there is a bug. 
On the other hand, Tm calculations using nearest neighbour thermodynamics are sensible to the input, especially when having mismatches:

- as you said, there is only litte data available for neighbouring mismatches (see also #682), I'm happy to include them if more are availabe.
- I wonder if these calculations can be used for multiple mismatches at all. I have to have a look in the paper if they ever included primers with more than one or two mismatches. If not, we could throw an error or a warning if more than one mismatch is present. But probably it's enough to write a warning into the documentation.

Finally, I don't think that any existing Tm calculation method gives reliable results for two nucleic acids that don't match (which is in principle what you are trying). These methods implicitely expect that the primer does nearly match. It's similar to trying to do an assembly with an aligment program.

**Comment by larssnip** [comment]
Yes, I think that a warning in the documentation is fine. I have been doing some testing, and found that if the two oligos have no more than 25% mismatches the computations are fine, in the sense they never seems to produce Extreme high temperatures. They produce largish negative values, but that is less of a concern. It just means there will be no hydridization...

In my code I test for percent mismatch first, and if it is above 25% the Tm is returned as 0.

I don't think the code itself has a bug. It is more on the validity of the results when we are faced With multiple mismatches. This is due to a lack of knowledge about the biochemistry here, not a coding error.
